s3/s3.cc

```diff
@@ -0,0 +1,625 @@
+};
+
+static
+std::pair<sstring, sstring> parse_header(const sstring& line) {
+    auto i = std::find(line.begin(), line.end(), ':');
+    if (i != line.end()) {
+        sstring key(line.begin(), i);
+        ++i;
+        // Skip space after ":"
+        if (i != line.end() && *i == ' ') {
+            ++i;
+        }
+        sstring value(i, line.end());
+        return std::make_pair(std::move(key), std::move(value));
+    } else {
+        return std::make_pair(line, sstring());
+    }
+}
+
+// Keep headers alive around async operation.
+static
+future<http_response> make_http_request(connection_ptr con,
+                                        const sstring& method,
+                                        const sstring& path,
+                                        const std::map<sstring, sstring>& headers,
+                                        std::optional<http_content_writer> content) {
+    log.debug("conn {}: {} {} {}", fmt::ptr(&*con), method, path, headers);
+
+    sstring method_ = method;
+    sstring path_ = path;
+    auto& out = con->out();
+
+    co_await out.write(method_);
+    co_await out.write(" ");
+    co_await out.write(path_);
+    co_await out.write(" HTTP/1.1\r\n");
+
+    for (auto&& [key, value] : headers) {
+        co_await out.write(key);
+        co_await out.write(": ");
+        co_await out.write(value);
+        co_await out.write("\r\n");
+    }
+
+    if (content) {
+        co_await out.write("Content-length: ");
+        co_await out.write(to_sstring(content->size));
+        co_await out.write("\r\n");
+        co_await out.write("\r\n");
+        co_await content->writer(out);
+    } else {
+        co_await out.write("\r\n");
+    }
+
+    co_await out.flush();
+
+    auto& in = con->in();
+    line_consumer lc;
+    co_await in.consume(lc);
+    auto status_line = lc.get();
+    log.debug("conn {}: status line: {}", fmt::ptr(&*con), status_line);
+
+    // Example status_line:
+    // HTTP/1.1 200 OK
+
+    // Parse response code
+    auto i = status_line.find(' ');
+    if (i == sstring::npos) {
+        throw std::runtime_error(format("invalid response line: {}", status_line));
+    }
+    ++i;
+    auto e = status_line.find(' ', i);
+    if (i == sstring::npos) {
+        throw std::runtime_error(format("invalid response line: {}", status_line));
+    }
+
+    auto response_code = std::atoi(status_line.substr(i, e - i).begin());
+
+    size_t content_length = 0;
+
+    // Parse response headers
+    std::unordered_map<sstring, sstring> response_headers;
+    while (1) {
+        co_await in.consume(lc);
+        auto line = lc.get();
+        if (line.empty()) {
+            break;
+        }
+        log.debug("conn {}: response header: {}", fmt::ptr(&*con), line);
+        auto hdr = parse_header(line);
+        if (hdr.first == "Content-Length") {
+            content_length = std::atoi(hdr.second.begin());
+        }
+        response_headers.emplace(std::move(hdr));
```

Comment:
We do (`http_response_parser`), I experimented with porting the code to this mechanism today, but I keep receiving very weird failures from `s3_client_test` and I haven't succeeded in making it work. From the logs and wireshark it looks like my new code based on `seastar::http_response_parser` for some reason fails to properly **upload** the test payload, which is done via http *request*, and not a *response*, so it's a little mind-blowing. In any case, I'll retry tomorrow, maybe I'll figure out what I did wrong (or maybe we have a bug in the response parser).

By the way, historically this parser couldn't be used during the hackathon, because it didn't parse the response code - it does that now.

Reply:
Done - the problem was with the fact that this code still sends the requests manually and some stray whitespace character ruined the communication. That's one more reason to seriously consider implementing a proper HTTP client for the sake of S3 integration (if we don't want to rely on AWS SDK alone, and we probably don't).